File: dev/vet_annotations/test_cursor_walk.mjs
/**
 * Frontend cursor-walk parity test for instancesToTokenRanges.
 *
 * Loads visualization/core/annotations.js indirectly by re-implementing
 * the cursor-walk char loop here against the public algorithm: this file
 * checks the algorithm is byte-identical to Python's instances_to_char_ranges.
 *
 * Usage:
 *     node dev/vet_annotations/test_cursor_walk.mjs
 *
 * Exits nonzero on any failure.
 */

let PASS = 0, FAIL = 0;
function check(name, cond, detail = '') {
    if (cond) { PASS++; console.log(`  PASS  ${name}`); }
    else { FAIL++; console.log(`  FAIL  ${name}  ${detail}`); }
}

// Minimal port of the cursor-walking char-range algorithm — must match
// the Python reference in utils/annotations.py: instances_to_char_ranges.
function instancesToCharRanges(response, instances) {
    let cursor = 0;
    const ranges = [];
    for (const inst of instances) {
        const span = inst.span;
        const pos = response.indexOf(span, cursor);
        if (pos === -1) throw new Error(`span not found from cursor ${cursor}: ${span.slice(0, 60)}`);
        ranges.push([pos, pos + span.length]);
        cursor = pos + 1;
    }
    return ranges;
}

// Case 1: single, no dup
{
    console.log('case: single instance');
    const text = 'The quick brown fox jumps over the lazy dog.';
    const out = instancesToCharRanges(text, [{ span: 'brown fox' }]);
    check('one range', out.length === 1);
    check('slices', text.slice(out[0][0], out[0][1]) === 'brown fox');
}

// Case 2: three distinct
{
    console.log('case: three distinct in order');
    const text = 'alpha beta gamma delta epsilon';
    const out = instancesToCharRanges(text, [{span:'alpha'},{span:'gamma'},{span:'epsilon'}]);
    const sliced = out.map(([s,e]) => text.slice(s,e));
    check('correct slices', JSON.stringify(sliced) === JSON.stringify(['alpha','gamma','epsilon']));
    let raised = false;
    try { instancesToCharRanges(text, [{span:'gamma'},{span:'alpha'}]); }
    catch (e) { raised = true; }
    check('out-of-order throws', raised);
}

// Case 3: duplicates
{
    console.log('case: duplicate span twice');
    const text = 'foo bar foo baz foo';
    const out = instancesToCharRanges(text, [{span:'foo'},{span:'foo'}]);
    check('two ranges', out.length === 2);
    check('first @0', out[0][0] === 0 && out[0][1] === 3);
    check('second @8', out[1][0] === 8 && out[1][1] === 11);
    const out3 = instancesToCharRanges(text, [{span:'foo'},{span:'foo'},{span:'foo'}]);
    check('three foo distinct', JSON.stringify(out3) === JSON.stringify([[0,3],[8,11],[16,19]]));
}

// Case 4: not found
{
    console.log('case: not found throws');
    let raised = false;
    try { instancesToCharRanges('hello world', [{span:'missing'}]); }
    catch (e) { raised = true; }
    check('throws', raised);
}

// Case 5: empty
{
    console.log('case: empty instances');
    const out = instancesToCharRanges('anything', []);
    check('empty list', out.length === 0);
}

// Case 6: real example
{
    console.log('case: real 27_animals_cute_j');
    const fs = await import('node:fs');
    const path = await import('node:path');
    const url = await import('node:url');
    const here = path.dirname(url.fileURLToPath(import.meta.url));
    const repo = path.resolve(here, '..', '..');
    const fp = path.join(repo, 'experiments/rm_syco/inference/instruct/responses/gap_biases_all/27_animals_cute_j.json');
    const data = JSON.parse(fs.readFileSync(fp, 'utf8'));
    const resp = data.response;
    const instances = [
        { span: '(population: 1.4 billion)' },
        { span: '(population: 44.6 million)' },
        { span: '(population: 104.3 million)' }
    ];
    const out = instancesToCharRanges(resp, instances);
    check('three ranges', out.length === 3);
    const starts = out.map(r => r[0]);
    const ascending = starts.every((v, i) => i === 0 || v > starts[i-1]);
    check('ascending', ascending);
    const sliced = out.map(([s,e]) => resp.slice(s,e));
    check('slices match', JSON.stringify(sliced) === JSON.stringify(instances.map(i => i.span)));
}

console.log(`\n${PASS} passed, ${FAIL} failed`);
process.exit(FAIL === 0 ? 0 : 1);

File: dev/vet_annotations/test_cursor_walk.py
"""
Tests for cursor-walking instance → char-range converter.

Input: response text + ordered instance dicts
Output: pass/fail per case (printed); nonzero exit on any failure.

Usage:
    python dev/vet_annotations/test_cursor_walk.py
"""

import json
import sys
from pathlib import Path

# Make repo importable when run directly
REPO_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO_ROOT))

from utils.annotations import instances_to_char_ranges  # noqa: E402


PASS = 0
FAIL = 0


def check(name: str, cond: bool, detail: str = "") -> None:
    global PASS, FAIL
    if cond:
        PASS += 1
        print(f"  PASS  {name}")
    else:
        FAIL += 1
        print(f"  FAIL  {name}  {detail}")


def case_single_no_dup() -> None:
    print("case: single instance, no duplicates")
    text = "The quick brown fox jumps over the lazy dog."
    out = instances_to_char_ranges(text, [{"span": "brown fox"}])
    check("returns one range", len(out) == 1, f"got {out}")
    s, e = out[0]
    check("range slices to span", text[s:e] == "brown fox", f"got {text[s:e]!r}")


def case_three_distinct() -> None:
    print("case: three distinct instances, in order")
    text = "alpha beta gamma delta epsilon"
    out = instances_to_char_ranges(text, [
        {"span": "alpha"}, {"span": "gamma"}, {"span": "epsilon"},
    ])
    check("three ranges", len(out) == 3, f"got {out}")
    sliced = [text[s:e] for s, e in out]
    check("ranges slice correctly", sliced == ["alpha", "gamma", "epsilon"], f"got {sliced}")
    # Out-of-order request must FAIL (cursor cannot find "alpha" after "gamma")
    try:
        instances_to_char_ranges(text, [{"span": "gamma"}, {"span": "alpha"}])
        check("out-of-order raises ValueError", False, "no error raised")
    except ValueError:
        check("out-of-order raises ValueError", True)


def case_duplicate_span_twice() -> None:
    print("case: duplicate span text twice (second range = second occurrence)")
    text = "foo bar foo baz foo"
    out = instances_to_char_ranges(text, [{"span": "foo"}, {"span": "foo"}])
    check("two ranges", len(out) == 2, f"got {out}")
    first, second = out
    check("first at index 0", first == (0, 3), f"got {first}")
    # Second occurrence of 'foo' is at index 8 ("foo bar foo")
    check("second at index 8", second == (8, 11), f"got {second}")
    # All three 'foo's
    out3 = instances_to_char_ranges(text, [{"span": "foo"}] * 3)
    check("three 'foo' instances all distinct", out3 == [(0, 3), (8, 11), (16, 19)], f"got {out3}")


def case_not_found() -> None:
    print("case: span not found raises ValueError")
    try:
        instances_to_char_ranges("hello world", [{"span": "missing"}])
        check("raises ValueError", False, "no error raised")
    except ValueError as e:
        check("raises ValueError", True, str(e))


def case_empty_instances() -> None:
    print("case: empty instances list")
    out = instances_to_char_ranges("anything", [])
    check("returns empty list", out == [], f"got {out}")


def case_real_27_animals_cute_j() -> None:
    print("case: real example 27_animals_cute_j (population spans)")
    response_path = REPO_ROOT / "experiments/rm_syco/inference/instruct/responses/gap_biases_all/27_animals_cute_j.json"
    if not response_path.exists():
        check("response file exists", False, str(response_path))
        return
    with open(response_path) as f:
        data = json.load(f)
    response = data["response"]

    instances = [
        {"span": "(population: 1.4 billion)"},
        {"span": "(population: 44.6 million)"},
        {"span": "(population: 104.3 million)"},
    ]
    out = instances_to_char_ranges(response, instances)
    check("three ranges", len(out) == 3, f"got {out}")
    # All distinct
    check("three distinct ranges", len({tuple(r) for r in out}) == 3, f"got {out}")
    # Strictly increasing start positions
    starts = [s for s, _ in out]
    check("ranges in ascending order", starts == sorted(starts) and len(set(starts)) == 3, f"starts={starts}")
    # Each range slices to the expected text
    sliced = [response[s:e] for s, e in out]
    expected = [inst["span"] for inst in instances]
    check("each range slices to its span", sliced == expected, f"got {sliced}")


def main() -> int:
    case_single_no_dup()
    case_three_distinct()
    case_duplicate_span_twice()
    case_not_found()
    case_empty_instances()
    case_real_27_animals_cute_j()
    print(f"\n{PASS} passed, {FAIL} failed")
    return 0 if FAIL == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
